### ISSUE TITLE
Named targetPort cause canal to crash

### ISSUE THREAD

**Issue by webts** [rancher/longhorn#57]
Try to deploy on Rancher/Server Beta 3 > Custom Nodes 
version: v1.8.9-rancher1

Using default config: https://raw.githubusercontent.com/rancher/longhorn/v0.2/deploy/longhorn.yaml
I get the following errors
```
kubectl -n longhorn-system logs longhorn-flexvolume-driver-deployer-74869d8788-wd2kd`
WARN[0030] Failed to detect flexvolume dir, fall back to default: %!(EXTRA *errors.withStack=cannot discover Flexvolume Dir: cannot reach node configURI /api/v1/proxy/nodes/do-node1/configz: Get https://10.43.0.1:443/api/v1/proxy/nodes/do-node1/configz: dial tcp 10.43.0.1:443: i/o timeout)
FATA[0060] Error deploying Flexvolume driver: Get https://10.43.0.1:443/apis/extensions/v1beta1/namespaces/longhorn-system/daemonsets/longhorn-flexvolume-driver: dial tcp 10.43.0.1:443: i/o timeout
```

Using modified config with FLEXVOLUME_DIR = "/var/lib/kubelet/volumeplugins" (path from `ps aux|grep kubelet`)
I get the following error 
```
kubectl -n longhorn-system logs longhorn-flexvolume-driver-deployer-77cffbcb74-849dj
FATA[0030] Error deploying Flexvolume driver: Get https://10.43.0.1:443/apis/extensions/v1beta1/namespaces/longhorn-system/daemonsets/longhorn-flexvolume-driver: dial tcp 10.43.0.1:443: i/o timeout
```

**Comment by yasker** [comment]
What's the Kubernetes distro on the nodes? Seems the deployer cannot reach the Kubernetes API server (which it thought running at `10.43.0.1:443`)

**Comment by yasker** [comment]
Reproduced, looking into it.

**Comment by yasker** [comment]
OK, a bug for Canal. Canal (Kubernetes Network plugin) crashed with a named port with dash (`-`) for Longhorn UI.

Run this against 0.2 yaml file to workaround it (and wait a while for canal to restart)
```
sed -i "s/targetPort:\ longhorn-ui/targetPort:\ 8000/g" longhorn.yaml
```

You can check the status of canal using
```
k -n kube-system get pod
```

Make sure `canal-xxxx` are all running.

**Comment by yasker** [comment]
Log for the crashed canal
```
kubelet logs -n kube-system canal-rr42s calico-node
```
```
2018-04-14 01:00:55.482 [INFO][208] felix.go 735: Reading from dataplane driver pipe...
2018-04-14 01:00:55.500 [INFO][208] felix.go 480: Health enabled.  Starting server. port=9099
2018-04-14 01:00:55.500 [INFO][208] health.go 144: Starting health endpoints port=9099
2018-04-14 01:00:55.501 [INFO][208] felix.go 570: No driver process to monitor
2018-04-14 01:00:55.502 [INFO][208] syncer.go 638: Syncing NetworkPolicy
2018-04-14 01:00:55.512 [PANIC][208] conversion.go 448: Invalid port %+v: %slonghorn-ui invalid port format (longhorn-ui)
panic: (*logrus.Entry) (0x1723500,0xc42012d540)

goroutine 62 [running]:
github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus.Entry.log(0xc420014cd0, 0xc420308030, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420014a00, ...)
    /go/src/github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus/entry.go:124 +0x340
github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus.(*Entry).Panic(0xc420014aa0, 0xc420601160, 0x3, 0x3)
    /go/src/github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus/entry.go:169 +0x11b
github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus.(*Logger).Panic(0xc420014cd0, 0xc420601160, 0x3, 0x3)
    /go/src/github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus/logger.go:235 +0x6e
github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus.Panic(0xc420601160, 0x3, 0x3)
    /go/src/github.com/projectcalico/felix/vendor/github.com/sirupsen/logrus/exported.go:107 +0x4b
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.Converter.k8sPortToCalico(0xc4202dba00, 0xc420333300, 0x3, 0x3, 0xc4202dba30)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/conversion.go:448 +0x150
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.Converter.k8sPortToCalicoFields(0xc4202db9f0, 0xc420601238, 0x0, 0x0, 0x1)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/conversion.go:391 +0x5b
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.Converter.k8sRuleToCalico(0x0, 0x0, 0x0, 0xc42041d300, 0x1, 0x10, 0xc4202da860, 0xf, 0xc42012d401, 0x0, ...)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/conversion.go:356 +0x7a9
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.Converter.NetworkPolicyToPolicy(0xc420602788, 0xc42046e030, 0xc420601fc0, 0xc42038520c)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/conversion.go:203 +0x262
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).performSnapshot(0xc42012dea0, 0xc4201d2310, 0xc420571548, 0x1)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:653 +0x3b8f
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).readFromKubernetesAPI(0xc42012dea0)
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:318 +0x378
created by github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).Start
    /go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:221 +0x3f
```

**Comment by yasker** [comment]
Fixed in cb142bc4b2fda0d6e7ba75b0ca6c6e48da2b9e45.
